js, es6: meta-data on atoms

dubek opened the pull request:
Support the (optional, soft-fail) meta-data on atoms (added to the guide).

Before the fix:

```
Mal [javascript]
user> (meta (with-meta (atom 7) {"a" 1}))
Error: clone of non-collection: atom
    at Object._clone (/home/dubek/world/mal/js/types.js:77:15)
    at Function.with_meta (/home/dubek/world/mal/js/core.js:153:25)
    at _EVAL (/home/dubek/world/mal/js/stepA_mal.js:143:22)
    at EVAL (/home/dubek/world/mal/js/stepA_mal.js:151:18)
    at /home/dubek/world/mal/js/stepA_mal.js:55:45
    at Array.map (<anonymous>)
    at eval_ast (/home/dubek/world/mal/js/stepA_mal.js:55:20)
    at _EVAL (/home/dubek/world/mal/js/stepA_mal.js:138:18)
    at EVAL (/home/dubek/world/mal/js/stepA_mal.js:151:18)
    at rep (/home/dubek/world/mal/js/stepA_mal.js:162:40)
```

After the fix:

```
Mal [javascript]
user> (meta (with-meta (atom 7) {"a" 1}))
{"a" 1}
```

(fixed similarly for the es6 implementation.)

kanaka commented:
Metadata isn't supported on Clojure atoms. That doesn't mean we can't support it in mal but is there a use-case that motivates it? There is also a question of what the behavior should be. Most objects are cloned when meta-data is added. It's a little weird to clone an atom because then you have two references that can diverge (i.e. they are different atoms but start out pointing to the same underlying object). But mutating an atom also doesn't seem like the right behavior either (i.e. `with-meta` would be a pseudo form of reset! just for the metadata part).

dubek commented:
not in front of PC now; I wonder why that final test was added in stepA.

On Tue, Dec 17, 2019, 17:12 Joel Martin <notifications@github.com> wrote:

> Metadata isn't supported on Clojure atoms. That doesn't mean we can't
> support it in mal but is there a use-case that motivates it? There is also
> a question of what the behavior should be. Most objects are cloned when
> meta-data is added. It's a little weird to clone an atom because then you
> have two references that can diverge (i.e. they are different atoms but
> start out pointing to the same underlying object). But mutating an atom
> also doesn't seem like the right behavior either (i.e. with-meta would be
> a pseudo form of reset! just for the metadata part).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kanaka/mal/pull/481?email_source=notifications&email_token=AAACH62CSYEL3GKNZTCLE7DQZDT7VA5CNFSM4J3ZXURKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHCWF3Y#issuecomment-566584047>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAACH66U4QQGLHD62NVLUETQZDT7VANCNFSM4J3ZXURA>
> .
>

kanaka commented:
I added it commented out in b079f51028571bc603b8d43761c29ff56273bffc and without thinking about it. I think I just added metadata tests for all non-scalar types. It got uncommented a few months ago as a soft test in 77057e69200e1bdd1e6e3704787906324c031420 and I didn't think about it carefully at the time. Also the comment about it not being support by Clojure was dropped which made it look more official. Anyways, you're proposal was the first time I really though about it in any depth. It should probably just be dropped as a test and idea. What do you think?

dubek commented:
yes, remove this test completely and close this PR without merging.

On Tue, Dec 17, 2019, 21:42 Joel Martin <notifications@github.com> wrote:

> I added it commented out in b079f51
> <https://github.com/kanaka/mal/commit/b079f51028571bc603b8d43761c29ff56273bffc>
> and without thinking about it. I think I just added metadata tests for all
> non-scalar types. It got uncommented a few months ago as a soft test in
> 77057e6
> <https://github.com/kanaka/mal/commit/77057e69200e1bdd1e6e3704787906324c031420>
> and I didn't think about it carefully at the time. Also the comment about
> it not being support by Clojure was dropped which made it look more
> official. Anyways, you're proposal was the first time I really though about
> it in any depth. It should probably just be dropped as a test and idea.
> What do you think?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kanaka/mal/pull/481?email_source=notifications&email_token=AAACH6YOGUA3CMQOCUDWBNDQZETTPA5CNFSM4J3ZXURKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHDXAJY#issuecomment-566718503>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAACH6ZMTINQGZZTIS7Q5BLQZETTPANCNFSM4J3ZXURA>
> .
>

kanaka commented:
Resolved in b75dc79.
